src/package-transpilation-registry.js

```diff
@@ -0,0 +1,151 @@
+  wrapTranspiler: function (transpiler) {
+    var self = this
+    return {
+      getCachePath: function (sourceCode, filePath) {
+        var spec = self.getPackageTranspilerSpecForFilePath(filePath)
+        if (spec) {
+          return self.getCachePath(sourceCode, filePath, spec)
+        }
+
+        return transpiler.getCachePath(sourceCode, filePath)
+      },
+
+      compile: function (sourceCode, filePath) {
+        var spec = self.getPackageTranspilerSpecForFilePath(filePath)
+        if (spec) {
+          return self.transpileWithPackageTranspiler(sourceCode, filePath, spec)
+        }
+
+        return transpiler.compile(sourceCode, filePath)
+      },
+
+      shouldCompile: function (sourceCode, filePath) {
+        if (self.transpilerPaths[filePath]) {
+          return false
+        }
+        var spec = self.getPackageTranspilerSpecForFilePath(filePath)
+        if (spec) {
+          return true
+        }
+
+        return transpiler.shouldCompile(sourceCode, filePath)
+      }
+    }
+  },
+
+  getPackageTranspilerSpecForFilePath: function (filePath) {
+    if (this.specByFilePath[filePath] !== undefined) return this.specByFilePath[filePath]
+
+    var config = null
+    var spec = null
+    var thisPath = filePath
+    var lastPath = null
+    // Iterate parents from the file path to the root, checking at each level
+    // to see if a package manages transpilation for that directory.
+    // This means searching for a config for `/path/to/file/here.js` only
+    // only iterates four times, even if there are hundreds of configs registered.
+    while (thisPath !== lastPath) { // until we reach the root
+      if (config = this.configByPackagePath[thisPath]) { // eslint-disable-line no-cond-assign
+        for (var i = 0; i < config.specs.length; i++) {
+          spec = config.specs[i]
+          if (minimatch(filePath, path.join(config.path, spec.glob))) {
+            spec._config = config
+            this.specByFilePath[filePath] = spec
+            return spec
+          }
+        }
+      }
+
+      lastPath = thisPath
+      thisPath = path.resolve(thisPath, '..')
+    }
+
+    this.specByFilePath[filePath] = null
+    return null
+  },
+
+  getCachePath: function (sourceCode, filePath, spec) {
+    var transpilerPath = this.getTranspilerPath(spec)
+    var transpilerSource = spec._transpilerSource || fs.readFileSync(transpilerPath, 'utf8')
+    spec._transpilerSource = transpilerSource
+    var transpiler = this.getTranspiler(spec)
+
+    var hash = crypto
+      .createHash('sha1')
+      .update(JSON.stringify(spec.options || {}))
+      .update(transpilerSource, 'utf8')
+      .update(sourceCode, 'utf8')
+
+    var additionalCacheData
+    if (transpiler && transpiler.getCacheKeyData) {
+      additionalCacheData = transpiler.getCacheKeyData(sourceCode, filePath, spec.options)
+      hash.update(additionalCacheData, 'utf8')
+    }
+
+    return path.join('package-transpile', hash.digest('hex'))
```

Comment:
I wonder if it's worth grouping transpiled sources in their own subfolder named after the package... that might make it easier to delete them when debugging or something. I don't have a super strong opinion on this one but I thought I'd raise the idea.